### PR TITLE
chore(sinks): Part 1 of `HttpSink` refactor

### PR DESCRIPTION
**Pull request by LucioFranco**
This PR is the first PR int he process of finishing off the `HttpSink` and general sink based http refactoring. This sink only touches `clickhouse`, `elasticsearch` and `datadog_metrics` but internally simplifies things.

Related to #1659

### CODE REVIEW

File: src/sinks/elasticsearch.rs [hunk 1]
```diff
@@ -122,6 +136,99 @@ enum ParseError {
...
+                "_type": self.doc_type,
+            }
+        });
+        maybe_set_id(
+            self.config.id_key.as_ref(),
+            action.pointer_mut("/index").unwrap(),
+            &event,
+        );
+
+        println!("event: {}", action);
```
Comment (LucioFranco) [inline src/sinks/elasticsearch.rs]:
```suggestion
```

File: src/sinks/elasticsearch.rs [hunk 2]
```diff
@@ -122,6 +136,99 @@ enum ParseError {
...
+                for (header, value) in headers {
+                    builder.header(&header[..], &value[..]);
+                }
+            }
+
+            if let Some(auth) = &self.authorization {
+                builder.header("Authorization", &auth[..]);
+            }
+
+            println!("REQUEST len: {}", events.len());
```
Comment (LucioFranco) [inline src/sinks/elasticsearch.rs]:
```suggestion
```

File: src/sinks/elasticsearch.rs [hunk 3]
```diff
@@ -576,6 +564,7 @@ mod integration_tests {
             .json::<elastic_responses::search::SearchResponse<Value>>()
             .unwrap();
 
+        println!("response {:?}", response);
```
Comment (LucioFranco) [inline src/sinks/elasticsearch.rs]:
```suggestion
```

File: src/sinks/util/batch.rs [hunk 4]
```diff
@@ -102,6 +103,7 @@ where
     }
 
     pub fn from_settings(inner: S, batch: B, settings: BatchSettings) -> Self {
+        println!("settings: {:?}", settings);
```
Comment (LucioFranco) [inline src/sinks/util/batch.rs]:
```suggestion
```

**Comment by Hoverbear** [comment]
Ha ha

File: src/sinks/datadog_metrics.rs [hunk 5]
```diff
@@ -104,96 +109,78 @@ inventory::submit! {
...
-    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
-        self.inner.poll_ready()
-    }
+    let request = hyper::Request::get(uri).body(hyper::Body::empty()).unwrap();
 
-    fn call(&mut self, items: Vec<Metric>) -> Self::Future {
-        let now = Utc::now().timestamp();
-        let interval = now - self.state.last_sent_timestamp;
-        self.state.last_sent_timestamp = now;
+    let https = HttpsConnector::new(4).expect("TLS initialization failed");
```
Comment (bruceg) [inline src/sinks/datadog_metrics.rs]: I keep wondering, do we really need 4 threads for a health check?

File: src/sinks/util/http.rs [hunk 6]
```diff
@@ -90,8 +90,14 @@ where
     type SinkError = ();
 
     fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        if self.slot.is_some() {
+            self.poll_complete()?;
+            return Ok(AsyncSink::NotReady(item));
+        }
```
Comment (bruceg) [inline src/sinks/util/http.rs]: I had to read this three times (and the struct definition) before I clued in what was going on here. Some commentary about the `slot` handling here and below would probably be appropriate.

File: src/sinks/datadog_metrics.rs (hunk 5 above)
Comment (LucioFranco) [reply src/sinks/datadog_metrics.rs]: This should be `HttpClient` anyways.

File: src/sinks/util/http.rs (hunk 6 above)
Comment (LucioFranco) [reply src/sinks/util/http.rs]: I added some more docs to the struct, but the `slot` item on the struct has a comment and links to the `with_flat_map` contaminator as well where this trick is handled.

**Pull request merged**